kube-client/src/api/subresource.rs

```diff
@@ -128,6 +128,178 @@ where
+    ///     let client = Client::try_default().await?;
+    ///     let pods: Api<Pod> = Api::namespaced(client, "apps");
+    ///     let pp = PostParams::default();
+    ///
+    ///     // Get pod object with ephemeral containers.
+    ///     let mut mypod = pods.get_ephemeral_containers("mypod").await?;
+    ///     
+    ///     // If there were existing ephemeral containers, we would have to append
+    ///     // new containers to the list before calling replace_ephemeral_containers.
+    ///     assert_eq(mypod.spec.ephemeral_containers, None);
+    ///
+    ///     // Add an ephemeral container to the pod object.
+    ///     mypod.spec.ephemeral_containers = Some(serde_json::from_value(serde_json::json!([
+    ///        {
+    ///            "name": "myephemeralcontainer",
+    ///            "image": "busybox:1.34.1",
+    ///            "command": ["sh", "-c", "sleep 20"],
+    ///        },
+    ///     ]))?);
+    ///
+    ///     pods.replace_ephemeral_containers("mypod", &pp, serde_json::to_vec(&o)?).await?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn replace_ephemeral_containers(
+        &self,
+        name: &str,
+        pp: &PostParams,
+        data: Vec<u8>,
+    ) -> Result<K> {
+        let mut req = self
+            .request
+            .replace_subresource("ephemeralcontainers", name, pp, data)
+            .map_err(Error::BuildRequest)?;
+        req.extensions_mut().insert("replace_ephemeralcontainers");
+        self.client.request::<K>(req).await
+    }
+
+    /// Patch the `EphemeralContainers` sub resource
+    ///
+    /// If an invalid patch is provided the method will **not** always return
+    /// an error and no changes will be made to the target object in
+    /// the cluster. See the examples below.
+    ///
+    /// Note that any partial object containing the ephemeral containers
+    /// sub resource is valid as long as the complete structure
+    /// for the object is present, as shown below.
+    ///
+    /// See the `Kubernetes` [documentation](https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/)
+    /// for more information about ephemeral containers.
+    ///
+    /// **Valid and invalid Examples for patching a pod**
+    ///
+    /// Valid as this can be merged with a complete Pod object:
+    /// ```rust,no_run
+    ///     let patch = serde_json::json!({
+    ///        "spec":{
+    ///        "ephemeralContainers": [
+    ///        {
+    ///            "name": "myephemeralcontainer",
+    ///            "image": "busybox:1.34.1",
+    ///            "command": ["sh", "-c", "sleep 20"],
+    ///        },
+    ///        ]
+    ///     }});
+    ///
+    /// ```
+    ///
+    /// Not valid as the outer layer of the `Pod` object is missing, instead
+    /// we have provided a partial `Pod` spec. Again note that no error will
+    /// be returned, the patch will simply have no effect.
```

Comment:
After a quick investigation I think this is caused by a Pod object happily being deserialized from a PodSpec json snippet. Where we actually end up sending an empty Pod object as the patch. I will perform more tests later, but essentially what I think happens is: 
1. we pass in a `Patch<P>`, where P is a PodSpec: https://github.com/kube-rs/kube/blob/4594a4f60b903b43d2f1873c2c353de24822dc25/kube-client/src/api/subresource.rs#L278
2. a request is built and the pod spec is serialized as the body, at this point we are stilling dealing with P: https://github.com/kube-rs/kube/blob/fdd799d4230aa09405881f18b946656637fed762/kube-core/src/request.rs#L259
3. Then before the actual request is performed the body is mapped to the return type of method which is Pod in this case, with this line, where I think we end up producing an empty Pod object as the body: https://github.com/kube-rs/kube/blob/75578903318a16688e14e30afa8745eee2c12f95/kube-client/src/client/mod.rs#L225.

Does this make sense? :thinking:

Reply:
hm, that should be easily testable with a println on the result there, but that seems unlikely to me.
the Request fn is like line-for-line equivalent to the `patch_metadata`.

i'm going to backtrack a bit here and say that this might not be inconsistent and "that bad" after all. types already need the wrapping in general, but we don't notice it for things like `status` because it's top level. noticed that it's also required for `patch_metadata` yesterday.

thus I think it's fine for the example to fully wrap in the json! patch:

```rust
serde_json::from_value(serde_json::json!({
    ///     "apiVersion": "v1",
    ///     "kind": "Pod",
    ///     "spec": {
    ///        "ephemeralContainers": [
    ///        {
    ///            "name": "myephemeralcontainer",
    ///            "image": "busybox:1.34.1",
    ///            "command": ["sh", "-c", "sleep 20"],
    ///        },
    ///     }
    /// }))?;
```

and just use server side apply in the example via `Patch::Apply`. If it's actually the same as `patch_metadata`, then you can omit the typemeta if you use a `Patch::Merge` (but including typemeta is valid in both cases)

Reply:
> After a quick investigation I think this is caused by a Pod object happily being deserialized from a PodSpec json snippet. Where we actually end up sending an empty Pod object as the patch. I will perform more tests later, but essentially what I think happens is:

So my hypothesis was incorrect haha, I am not sure why no error is returned exactly other than perhaps Kubernetes ignoring all of the fields that do not belong to the resource being patched.

Just to be clear this is the case I am describing in the docs and brought up earlier.

```rust
        let patch: PodSpec = serde_json::from_value(json!({
           "spec" {
                 "ephemeralContainers:" [{
                    "name": "myephemeralcontainer2",
                    "image": "busybox:1.35.0",
                    "command": ["sh", "-c", "sleep 1"],
            }]
        }))?;

       pods.patch_ephemeral_containers(
                &pod_name,
                &PatchParams::default(),
                &Patch::Strategic(patch),
            )
            .await?
            .spec
            .unwrap()
            .ephemeral_containers
            .expect("could find ephemeral container");
```

No error will be returned here so it would be easy to run this and think everything is fine.

>i'm going to backtrack a bit here and say that this might not be inconsistent and "that bad" after all. types already need the wrapping in general, but we don't notice it for things like status because it's top level (and maybe also because we were using merge patches were it genuinely is not required).

I think the wrapping is just fine and consistent with how the rest of kube-rs and how client-go behaves as well. I am only worried about the lack of errors when submitting an invalid patch, hence the examples in the documentation. If you would prefer to just focus on the working example in the documentation and drop the valid/invalid comparisons I will remove them.

> and just use server side apply in the example via Patch::Apply. If it's actually the same as patch_metadata, then you can omit the typemeta if you use a Patch::Merge (but including typemeta is valid in both cases)

The example currently uses `Patch::Strategic` and no type metadata is required.

Reply:
> perhaps Kubernetes ignoring all of the fields that do not belong to the resource being patched.

mm, that's what i am thinking as well.

> I am only worried about the lack of errors when submitting an invalid patch, hence the examples in the documentation.

yeah, i definitely want to have users having clear info about when strange things can fail as well. just wanted to highlight two separate concerns here;

1. certain patch types seem to need typemeta included in the patch as a general thing
2. certain api endpoints not giving you back errors for invalid patches.

it is useful to highlight awkward patch behaviour under the ephemeral resources, but we should document general findings under `Patch` docs so people can more easily notice it. we should possibly even extend `PatchParams::validate` for this if it's a footgun (we can make a separate issue for this though).

if the broken api endpoint is just the ephemeral container one, (e.g. if it cannot be reproduced via straight `Api::patch`), then it makes sense to have an extra warning for these methods. it also might be worth raising a bug upstream against kubernetes long term.

Reply:
> 
> if the broken api endpoint is just the ephemeral container one, (e.g. if it cannot be reproduced via straight `Api::patch`), then it makes sense to have an extra warning for these methods. it also might be worth raising a bug upstream against kubernetes long term.

I just tried this with `Api::patch` and I get the same "bad" behavior :(. I will move the comments regarding this to the general `Patch` docs and make a separate issue to discuss  extending `PatchParams::validate`.

> certain patch types seem to need typemeta included in the patch as a general thing

Ah I see, is there anything I should address related to this?

Reply:
I might have overdone the `Patch` docs haha, let me know what you think.

Reply:
`PatchParams::validate_strict` seems to already do what we need to cover for this case. I tried it out and it does reject the invalid patch example :partying_face: